Preserve strides of custom Triton kernel args

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116219

Summary: Currently, we [`clone`](https://github.com/pytorch/pytorch/blob/19207b9183baced683d675c341b7da051cd16440/torch/_inductor/lowering.py#L5273) every `TensorBox` argument of custom Triton kernels while lowering them to the Inductor IR, during which the stride information of the kernel inputs is lost. This is problematic in the common case when the strides of a `torch.Tensor` argument are passed as scalars to a custom Triton kernel alongside the tensor itself (due to the underlying Triton code interpreting the tensors as raw pointers, so the contained stride semantics of the `torch.Tensor` is lost).

In this PR, we add an extended version of the existing [`clone` lowering](https://github.com/pytorch/pytorch/blob/19207b9183baced683d675c341b7da051cd16440/torch/_inductor/lowering.py#L2289)---`clone_preserve_reinterpret_view`---which carries over the `ir.ReinterpretVew` layers (if any) from the source `TensorBox` to the cloned one. The rationale behind adding a new function (and switching to it in the `triton_kernel_wrap` only for now) as opposed to extending the existing `clone` is keeping the semantics of the latter untouched, as it is a lowering of `torch.clone` (albeit incomplete, as the `memory_format` is currently ignored). Changing the existing `clone` would change the semantics which is not necessarily desirable in general. Open to suggestions, though.

Test Plan:

```
$ python test/dynamo/test_functions.py -k test_triton_kernel_strided_input
...
----------------------------------------------------------------------
Ran 1 test in 5.568s

OK
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler

aakhundov committed:
Preserve strides of custom Triton kenrel args

Summary: Currently, we [`clone`](https://github.com/pytorch/pytorch/blob/19207b9183baced683d675c341b7da051cd16440/torch/_inductor/lowering.py#L5273) every `TensorBox` argument of custom Triton kernels during lowering to the Inductor IR, during which the stride information is lost. This is problematic for the common case when the strides of a `torch.Tensor` argument are passed as scalars to a custom Triton kenrel alongside the tensor itself (due to the underlying Triton code interpreting the tensors as row pointers, so the contained stride semantics of `torch.Tensor` is lost).

In this PR, we add another version of `clone`---`clone_preserve_reinterpret_view`---which does carries over the `ir.ReinterpretVew` layers from the source `TensorBoxed` to the cloned one. The rationale of adding a new function (and switching to it in the `triton_kernel_wrap` only for now) is keeping the semantics of the existing `clone` untouched, as it is a lowering of `torch.clone` (albeit incomplete, as the `memory_format` is currently ignored): changing the exiting `clone` would change the semantics which is not necessarily desirable in general. Open to suggestions, though.

Test Plan:

```
$ python test/dynamo/test_functions.py -k test_triton_kernel_strided_input
...
----------------------------------------------------------------------
Ran 1 test in 5.568s

OK

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/ir.py:
```diff
@@ -3681,7 +3681,7 @@ def realize_input(cls, x):
             return cls.realize_input(x.data)
         if isinstance(x, ReinterpretView):
-            return x
+            return cls.realize_input(x.data)
```
Hmm, does this need to re-run the view off of the realized base? I.e. `ReinterpretView(cls.realize_input(x.data), x.layout)`

aakhundov replied:
So my understanding is that realization is happening in-place. But the function returns something, so what you're suggesting sounds required. Let me look into it.

bdhirsh commented on test/dynamo/test_functions.py:
```diff
@@ -2427,6 +2427,29 @@ def f(x):
+    @requires_cuda()
+    def test_triton_kernel_strided_input(self):
+        def f(inp):
+            left, right = torch.split(inp, [128, 128], dim=1)
```
Do you mind also adding a test where `right` gets used as the input to the triton kernel? That will exercise a version where we're passing a tensor with a non-zero storage offset to triton.

I think... we shouldn't need to to preserve the storage_offset when we clone, but it would be good to have a test for it.

aakhundov replied:
Sure, will do!

bdhirsh commented:
Deferring to Oguz/Jason for the rest of the review

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -2296,6 +2296,26 @@ def clone(x, *, memory_format=None):
     )
 
 
+def clone_preserve_reinterpret_view(x):
```
I know there are some other clone while preserving strides type functions floating around, might be worth a checking to see if we could consolidate any of them.

aakhundov committed:
Update on "Preserve strides of custom Triton kernel args"


Summary: Currently, we [`clone`](https://github.com/pytorch/pytorch/blob/19207b9183baced683d675c341b7da051cd16440/torch/_inductor/lowering.py#L5273) every `TensorBox` argument of custom Triton kernels while lowering them to the Inductor IR, during which the stride information of the kernel inputs is lost. This is problematic in the common case when the strides of a `torch.Tensor` argument are passed as scalars to a custom Triton kernel alongside the tensor itself (due to the underlying Triton code interpreting the tensors as raw pointers, so the contained stride semantics of the `torch.Tensor` is lost).

In this PR, we add an extended version of the existing [`clone` lowering](https://github.com/pytorch/pytorch/blob/19207b9183baced683d675c341b7da051cd16440/torch/_inductor/lowering.py#L2289)---`clone_preserve_reinterpret_view`---which carries over the `ir.ReinterpretVew` layers (if any) from the source `TensorBox` to the cloned one. The rationale behind adding a new function (and switching to it in the `triton_kernel_wrap` only for now) as opposed to extending the existing `clone` is keeping the semantics of the latter untouched, as it is a lowering of `torch.clone` (albeit incomplete, as the `memory_format` is currently ignored). Changing the existing `clone` would change the semantics which is not necessarily desirable in general. Open to suggestions, though.

Test Plan:

```
$ python test/dynamo/test_functions.py -k test_triton_kernel_strided_input
...
----------------------------------------------------------------------
Ran 1 test in 5.568s

OK
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge

aakhundov commented:
@pytorchbot merge
